### ISSUE TITLE
Make illegal states of `Block<T>` & `Transaction<T>` unrepresentable

### ISSUE THREAD

**Issue by dahlia** [planetarium/libplanet#1164]
Related: #1146.

To be elaborated.

**Comment by greymistcube** [comment]
# Preamble

Well, this is really getting out of hand.

Indeed, `Block<T>` is being used in multiple stages without regards to its validity. There are historical reasons why this has come to be the case, but it might be preferable to overhaul the whole class at this point.

After a careful inspection, there seems to be two major issues:
* The content of `Block<T>` is largely divided into two groups.
  * Those not dependent on `PreEvaluationHash` and those dependent on `PreEvaluationHash`.
* Evaluation process is rather convoluted.

To summarize the issues even further, in essence, the design of `Block<T>` seems to have evolved far too much for its surrounding support classes/methods, which are way too generic, to be useful. I suggest we decide asap whether to systematically support "two-stage", or even "multi-stage", `Block<T>`.

There are also policy level complications arising from having `StateRootHash` and `Hash` derived from `PreEvaluationHash`, due to both not being secured by PoW for the newest block. Although this is not entirely a separate issue, these should be discussed more in depth if necessary.

For everyone's sanity, and especially mine, some overview of the creation processes of a `Block<T>` might be in order.

# Creation of `Block<T>`

There are mainly two ways for a `Block<T>` instance to be created: mining and deserialization.

## Mining

This is where a `Block<T>` gets created through mining. In summary, `Block<T>()` constructor is called with appropriate `Nonce` value. However, the actual process is much more complicated.

* For the creation of a new `Block<T>` through mining, `Block<T>.Mine()` is called from `BlockChain<T>.MineBlock()`.
* `Block<T>.Mine()` returns a new `Block<T>` instance only containing minimal amount of information without either `StateRootHash` or `Hash`.
  * `Transaction<T>`'s are ordered by `Id` *before* `Block<T>()` is called.
  * Inside `Block<T>()`, `Transaction<T>` is ordered by `Id` once again in preparation for calculating `PreEvaluationHash` on the fly.
    * There is an error where `SerializeForHash()` is called to *include* `StateRootHash` when calculating `PreEvaluationHash`. Since `StateRootHash` has not yet been assigned, `Block<T>.StateRootHash` of `null` gets converted to `BlockHeader.StateRootHash` of `Empty`, and then to non-existance in its dictionary representation, resulting in accidental proper evaluation of `PreEvaluationHash`.
  * Once `PreEvaluationHash` is calculated in `Block<T>()`, `StateRootHash` is assigned `null` by default, and `Block<T>.Hash` is calculated via calling `SerializeForHash()` again, which results in having the same value as `PreEvaluationHash`.
    * Meanwhile, `Block<T>.Header` is called *twice* to create `BlockHeader` on the fly.
  * Once `PreEvaluationHash`, "`StateRootHash`", and "`Hash`" are assigned, `Transaction<T>`s are reordered to its execution order, which depend on `PreEvaluationHash` value as random seed.
    * This is a rather heavy operation, or so I have heard, so should be used with caution.
* Once `Block<T>.Mine()` has returned such partial `Block<T>` instance to `BlockChain<T>.MineBlock()`, local `IReadOnlyList<ActionEvaluation>` is generated by calling `BlockEvaluator.EvaluateActions()` with said `Block<T>` and a `StateCompleterSet<T>` in preparation for getting the `StateRootHash` of the `Block<T>`.
  * With some `null` check, `StateCompleterSet<T>` is split to `AccountStateGetter` and `AccountBalanceGetter` and used to execute every `IAction` in the block via `Block<T>.Evaluate()`.
    * Inside `Block<T>.Evaluate()`, there is a *redundant* `null` check for `AccountStateGetter` and `AccountBalanceGetter`.
    * `Block<T>.Header` is then *validated*. Aside from on the fly property being called yet again, this does not seem to be the best place to validate the `Header` of a `Block<T>`.
        * Once `Block<T>.Header` is validated, `IAction`s in each `Transaction<T>` is evaluated by calling `EvalutateActionsPerTx()`.
        * *Yet another* `null` check for `AccountStateGetter` and `AccountBalanceGetter` is performed.
        * For each `Transaction<T>`, which is in evaluation order, even though this is all happening inside a *"partially"* constructed `Block<T>` instance, `Transaction<T>.EvaluateActionsGradually()` is called to evaluate `IAction`s inside a `Transaction<T>`.
            * Inside `Transaction<T>.EvaluateActionsGradually()`, everything gets passed on to `ActionEvaluation.EvaluateActionsGradually()`, which is a method from some external evaluation helper. This completes the cycle of going from an external helper method of `BlockEvaluator` to an internal method of `Block<T>` then to `Transaction<T>` then finally to an external helper method of `ActionEvaluation`.
  * Once `Block<T>.Evaluate()` is complete, `Block<T>` as an action is evaluated via `BlockEvaluator.EvaluateBlockAction()`.
* Now that necessary `IReadOnlyList<ActionEvaluation>` is generated, `BlockChain<T>.SetStates()` is called to obtain `StateRootHash` for the said `Block<T>` instance.
  * Inside `BlockChain<T>.SetStates()`, this is simply redirected to `StateStore.SetStates()`.
    * To calculate `StateRootHash`, `TrieStateStore.EvalState()` is called. Note `EvalState()` is *not* part of `IStateStore` interface.
    * Evaluated `StateRootHash` is stored with `Block<T>.Hash` which is the same as `PreEvaluationHash` at this point.
* Once `StateRootHash` is finally obtained, a new `Block<T>` instance is created from the old `Block<T>` and this `StateRootHash`.
  * The same constructor is called, only now with `StateRootHash` provided. The entire process of initial `Block<T>` creation above is repeated.
* With newly created `Block<T>`, `StateStore.SetStates()` is called once again to set evaluated `Block<T>` states with a proper `Hash`.

## Deserialization

This is when data is received over the network and the corresponding `Block<T>` is instantiated from the serialized data. The process is much more straightforward since all the necessary values are already provided. Note that the deserialization process does not share any usage of a constructor with the mining process. This has left a gap in validation where the receiving node blindly trusts the transaction order transmitted by another node.
